jurity/recommenders/interlist_diversity.py

```diff
@@ -0,0 +1,225 @@
+# -*- coding: utf-8 -*-
+# Copyright FMR LLC <opensource@fidelity.com>
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import List, Union, Tuple
+
+import numpy as np
+import pandas as pd
+import scipy.sparse as sp
+from sklearn.metrics.pairwise import cosine_similarity
+import warnings
+
+from jurity.utils import Constants
+
+
+def sample_users(df: pd.DataFrame, user_id_column: str = Constants.user_id,
+                 sample_size: float = None, seed: int = Constants.default_seed) -> pd.DataFrame:
+    """
+    Samples input data frame by selecting a random sample of users.
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        Data frame with a user_id_col column.
+    user_id_column: str
+        User id column name.
+    sample_size: float
+        Proportion of users to randomly sample for evaluation.
+    seed : int, default=Constants.default_seed
+        The seed used to create random state.
+
+    Returns
+    -------
+    Sampled data frame
+    """
+    rng = np.random.RandomState(seed)
+    users = df[user_id_column].unique()
+    users = rng.choice(users, size=int(len(users) * sample_size), replace=False)
+    return df[df[user_id_column].isin(users)]
+
+
+def tocsr(df: pd.DataFrame, user_id_column: str = Constants.user_id, item_id_column: str = Constants.item_id):
+    """
+    Transform data frame with user_id and item_id columns to sparse csr matrix.
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        Data frame with (user_id, item_id) in each row.
+    user_id_column: str
+        User id column name.
+    item_id_column: str
+        Item id column name.
+
+    Returns
+    -------
+    Sparse matrix with user-item interactions/recommendations.
+    """
+
+    # Map each user_id to (0, n_users)
+    users = list(df[user_id_column].unique())
+    user_map = dict(zip(users, range(len(users))))
+
+    # Map each user_id to (0, n_items)
+    items = list(df[item_id_column].unique())
+    item_map = dict(zip(items, range(len(items))))
+
+    # Update user_id, item_id values
+    user_ids = df[user_id_column].map(user_map).values
+    item_ids = df[item_id_column].map(item_map).values
+
+    # Convert to csr matrix
+    csr_matrix = sp.coo_matrix((np.ones(len(user_ids)), (user_ids, item_ids))).tocsr()
+
+    return csr_matrix
+
+
+def interlist_diversity(predicted_results: pd.DataFrame, click_column: str, k: int,
+                        user_id_column: str = Constants.user_id, item_id_column: str = Constants.item_id,
+                        sample_size: float = None, seed: int = Constants.default_seed,
+                        chunk_size: int = 1000) -> Tuple[float, int]:
+    """
+        Calculate inter-list diversity metric:
+
+            Inter-list-diversity = 1 - average(cosine_similarity(R_{u_i}, R_{u_j})),
+            where R_{u_i} is the binary indicator vector representing provided recommendations for user u_i, i \ne j.
+
+        Parameters
```

Comment:
curiosity question:
For a given user u, what's the relationship between the number of rows we have u_item1, u_item2, etc.. and the number k? 

Can we have fewer rows than k? 
Should all users have all items captured as rows?

Reply:
There can be 1 or more rows for any user in the set, fewer than `k` is allowed. We don't make too many assumptions about the input (it can be sorted/unsorted, have more or less rows than k, have wildly different # of recos per user, etc.). Jurity takes care of taking the top k rows per user (or less, if there aren't k rows) by sorting and grouping.

TLDR: We assume raw data and filter accordingly.

Reply:
When the answer you hope for arrives.. 🥳 👍